vis 0.4 (new formula)

esdf opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,54 @@
+    sha256 "149be31e0155c4694f77ea7264d9b398dd134eca0d00ff03358d91a6cfb2ea9d"
+  end
+
+  def script; <<-EOS.undent
```
Just inline this inside `def install` since it's only being used the once.

```ruby
(bin/"vise").write <<~EOS
   blah
EOS
```
etc.

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,54 @@
+
+  def script; <<-EOS.undent
+    #!/bin/sh
+    VIS_BASE=`brew --prefix vis`/libexec
```
Also:
```ruby
`brew --prefix vis`/libexec
```
Can be just:
```
#{libexec}
```

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,54 @@
+  sha256 "f11ba41cfb86dd39475960abfd12469de4da0ccfdb941f1d7680d89d987694c5"
+  head "https://github.com/martanne/vis.git"
+
+  depends_on "lua" => :build
```
This seems unlikely to be true? IIRC, installed Luarocks still require the `lua` interpreter around.

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,54 @@
+      end
+    end
+
+    # avoid inclusion of -lncursesw alongside -ltermkey because lncursesw might not exist (also avoid use of lncursesw altogether to simplify things)
```
Should probably be punted upstream.

DomT4 replied:
Ah, you already have in https://github.com/martanne/vis/issues/636. Thanks!

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,54 @@
+    (bin/"vise").write script
+  end
+
+  def caveats; <<-EOS.undent
```
`<<-EOS.undent` has been deprecated in favour of `<<~EOS`.

esdf commented:
@DomT4 Thanks for the comments! I have made the suggested changes

 - updated deprecated `<<-EOS.undent` with `<<~EOS`
 - inlined a `<<~EOS blah EOS`
 - replaced  ```
`brew --prefix vis\`/libexec
``` with `#{libexec}`
 - replaced `depends_on "lua" => :build` with `depends_on "lua"`

ilovezfs commented:
We can reconsider this once the ncurses issues are ironed out upstream. Thanks for the PR nonetheless @esdf!

esdf commented:
Ok great!

ilovezfs commented:
going to leave this open since we have https://github.com/martanne/vis/issues/636

esdf commented:
@ilovezfs The ncurses build issue is now fixed upstream https://github.com/martanne/vis/commit/73ef4885451655e518a1b30e621d8ba2d47c070d. Indeed if I change the Formula...
```ruby
  version "0.4"
  url "https://github.com/martanne/vis/archive/73ef4885451655e518a1b30e621d8ba2d47c070d.tar.gz"
  sha256 "90e166a28af7349ca40dc62ba8547453e44a8ae9de8dcc614d6c57e7f7dd01cd"
```
...then I can remove the `inreplace s.gsub` patching and can build&run fine. But then the Formula is not pointing at the 0.4 commit. Should do this, wait for a version bump upstream, point the formula at `head` or something else?

(I also looked at 
https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#formulae-revisions but I don't think that applies for a new formula)

ilovezfs commented:
```
  stable do
    url "https://github.com/martanne/vis/archive/v0.4.tar.gz"
    sha256 "f11ba41cfb86dd39475960abfd12469de4da0ccfdb941f1d7680d89d987694c5"

    patch do
      url "https://github.com/martanne/vis/commit/73ef4885.patch?full_index=1"
      sha256 "639f2b5ff708327c5d6b5b7d198b24cb4a71a72ae935a92bd765bd532c755603"
    end
  end
```

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  def caveats; <<~EOS
+    To avoid a name conflict with the OS X system utility /usr/bin/vis,
```
Minor nit: `macOS` these days, I guess.

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  test do
+    assert_match "vis v0.4 +curses +lua", shell_output("#{bin}/vise -v 2>&1")
```
And use `#{version}` here instead of hardcoding the version manually, although ILZ may well ask for a fuller functionality test if one is possible.

esdf commented:
@ilovezfs The patch works great, but when I come to do a brew audit I get this:
```
brew audit --new-formula vis
vis:
  * New formulae should not require patches to build. Patches should be submitted and accepted upstream first.
Error: 1 problem in 1 formula
```

DomT4 commented:
In this case it has been submitted & accepted upstream.

esdf commented:
I updated the Formula

 - includes the patch to https://github.com/martanne/vis/commit/73ef4885451655e518a1b30e621d8ba2d47c070d
 - removes the `inreplace`
 - `macOS` instead of `OS X`
 - `{#version}` in test instead of hardcoded version

esdf commented:
@DomT4 So it looks like the ci tests fail at the same place on all platforms
```
21:16:22 Error: 1 problem in 1 formula
21:16:22 vis:
21:16:22   * New formulae should not require patches to build. Patches should be submitted and accepted upstream first.
```
Is there something I should do to the Formula at this point?

DomT4 commented:
This may be temporarily blocked on me/ILZ/etc deciding what to do with `lua` in either https://github.com/Homebrew/homebrew-core/pull/21300 or https://github.com/Homebrew/homebrew-core/pull/21282.

esdf commented:
@DomT4 Exciting! I watch with anticipation.

DomT4 commented on Formula/vis.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def install
+    luapath = libexec/"vendor"
+    ENV["LUA_PATH"] = "#{luapath}/share/lua/5.2/?.lua"
```
The `5.2` references can now be changed to `5.3`.

esdf commented:
@DomT4 I have updated the formula to reference lua 5.3

esdf commented:
@DomT4 @ilovezfs Any update on the `lua` changes (#21300 or #21282)? Are there any changes needed for this formula?

DomT4 commented:
This is fine as-is in regards to the `lua` situation. I need to rebase & finish off https://github.com/Homebrew/homebrew-core/pull/21282 but that's not a blocker here. As for yay/nay-ing this PR, that be @ilovezfs' yard.

ilovezfs commented on Formula/vis.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  test do
+    assert_match "vis v#{version} +curses +lua", shell_output("#{bin}/vise -v 2>&1")
```
can we have a functional test here?

ilovezfs commented on Formula/vis.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def caveats; <<~EOS
+    To avoid a name conflict with the macOS system utility /usr/bin/vis,
+    this text editor must be invoked by calling `vise` ("vis-editor").
```
The wording seems a little awkward. Can we call `vise`, `vis-editor` or both (yes, I know, just `vise`)? Does "this text editor" refer to "/usr/bin/vis" or the formula (yes, I know, the formula)?

ilovezfs commented on Formula/vis.rb:
```diff
@@ -0,0 +1,53 @@
+    ENV["LUA_PATH"] = "#{luapath}/share/lua/5.3/?.lua"
+    ENV["LUA_CPATH"] = "#{luapath}/lib/lua/5.3/?.so"
+
+    resources.each do |r|
```
There's only one resource so use `resource("lpeg")`

ilovezfs commented on Formula/vis.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+  end
+
+  depends_on "lua"
```
alphabetize

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented:
Thanks for your first contribution to Homebrew, @esdf! Without people like you submitting PRs we couldn't run this project. You rock!
